common/locks/condition_variable_impl.go

```diff
@@ -0,0 +1,90 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package locks
+
+import (
+	"sync"
+)
+
+type (
+	ConditionVariableImpl struct {
+		lock Locker
+
+		chanLock sync.Mutex
+		channel  chan struct{}
+	}
+)
+
+var _ ConditionVariable = (*ConditionVariableImpl)(nil)
+
+func NewConditionVariable(
+	lock Locker,
+) *ConditionVariableImpl {
+	return &ConditionVariableImpl{
+		lock: lock,
+
+		chanLock: sync.Mutex{},
+		channel:  make(chan struct{}),
+	}
+}
+
+func (c *ConditionVariableImpl) Signal() {
+	c.chanLock.Lock()
+	defer c.chanLock.Unlock()
+
+	select {
+	case c.channel <- struct{}{}:
+	default:
+		//noop
+	}
+}
+
+func (c *ConditionVariableImpl) Broadcast() {
```

Comment:
SignalAll / NotifyAll?

Reply:
i am trying to mimic c / go's function names: 
https://pubs.opengroup.org/onlinepubs/007908799/xsh/pthread_cond_signal.html
https://golang.org/src/sync/cond.go?s=1867:1893#L63